### PR TITLE
reduce code protected by lock as much as possible

### PR DESCRIPTION
**Pull request by beniwohli**
this avoids deadlocks that are caused by code inside the lock
triggering log messages that could be captured and queued, which in turn
leads to a deadlock.

### CODE REVIEW

File: elasticapm/transport/base.py [hunk 1]
```diff
@@ -100,28 +98,28 @@ def flush(self, sync=False, start_flush_timer=True):
...
-                    try:
-                        self.send(data)
-                        self.handle_transport_success()
-                    except Exception as e:
-                        self.handle_transport_fail(e)
-            self._last_flush = timeit.default_timer()
-            if start_flush_timer:
-                self._start_flush_timer()
+        if queued_data and not self.state.should_try():
+            logger.error("dropping flushed data due to transport failure back-off")
```
Comment (graphaelli) [inline elasticapm/transport/base.py]: to be clear, this is the important line to move outside?

Comment (beniwohli) [reply elasticapm/transport/base.py]: that one, and the two `Starting/Cancelling flush timer`

**Comment by graphaelli** [comment]
For visibility - ran this branch with apm-integration-testing and it appears the issue remains.  That shouldn't stop this from being merged but we can't call it case closed.

**Comment by beniwohli** [comment]
Retracting this in favor of a more complete refactor of the transport queuing logic

**Pull request closed without merging**